src/sleep.rs

```diff
@@ -0,0 +1,376 @@
+#[derive(Debug)]
+pub struct UlpWakeup {}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl UlpWakeup {
+    pub fn apply(&self) -> Result<(), EspError> {
+        esp!(unsafe { esp_sleep_enable_ulp_wakeup() })?;
+        Ok(())
+    }
+}
+
+/// Struct for light sleep. Add wakeup sources to this struct, and then call sleep().
+#[derive(Debug, Default)]
+pub struct LightSleep<'a> {
+    pub timer: Option<TimerWakeup>,
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    pub rtc: Option<RtcWakeup<'a>>,
+    pub gpio: Option<GpioWakeup<'a>>,
+    pub uart: Option<UartWakeup<'a>>,
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    pub touch: Option<TouchWakeup>,
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    pub ulp: Option<UlpWakeup>,
+}
+
+impl<'a> LightSleep<'a> {
+    pub fn sleep(&self) -> Result<(), EspError> {
+        esp!(unsafe { esp_sleep_disable_wakeup_source(esp_sleep_source_t_ESP_SLEEP_WAKEUP_ALL) })?;
+
+        if let Some(timer) = &self.timer {
+            timer.apply()?;
+        }
+
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        if let Some(rtc) = &self.rtc {
+            rtc.apply()?;
+        }
+
+        if let Some(gpio) = &self.gpio {
+            gpio.apply()?;
+        }
+
+        if let Some(uart) = &self.uart {
+            uart.apply()?;
+        }
+
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        if let Some(touch) = &self.touch {
+            touch.apply()?;
+        }
+
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        if let Some(ulp) = &self.ulp {
+            ulp.apply()?;
+        }
+
+        esp!(unsafe { esp_light_sleep_start() })?;
+        Ok(())
+    }
+}
+
+/// Struct for deep sleep. Add wakeup sources to this struct, and then call sleep().
+#[derive(Debug, Default)]
+pub struct DeepSleep<'a> {
+    pub timer: Option<TimerWakeup>,
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    pub rtc: Option<RtcWakeup<'a>>,
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    pub touch: Option<TouchWakeup>,
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    pub ulp: Option<UlpWakeup>,
+    #[cfg(not(any(esp32, esp32s2, esp32s3)))]
+    pub _p: PhantomData<&'a ()>,
+}
+
+impl<'a> DeepSleep<'a> {
+    pub fn prepare(&self) -> Result<(), EspError> {
+        esp!(unsafe { esp_sleep_disable_wakeup_source(esp_sleep_source_t_ESP_SLEEP_WAKEUP_ALL) })?;
+
+        if let Some(timer) = &self.timer {
+            timer.apply()?;
+        }
+
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        if let Some(rtc) = &self.rtc {
+            rtc.apply()?;
+        }
+
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        if let Some(touch) = &self.touch {
+            touch.apply()?;
+        }
+
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        if let Some(ulp) = &self.ulp {
+            ulp.apply()?;
+        }
+        Ok(())
+    }
+
+    pub fn sleep(&self) -> ! {
+        unsafe { esp_deep_sleep_start() };
+
+        // Ensure that funciton never returns for esp idf version 5.x
+        #[cfg(esp_idf_version_major = "5")]
```

Comment:
I got different build results with this from 4 vs 5. Related to how `esp_deep_sleep_start()` is defined perhaps?

Will double check, but right now im not able to build 5.x, so needs to figure out whats wrong with my setup

Reply:
Checked the esp idf code for 4.4 vs 5.1. In 4.4 `esp_deep_sleep_start()` always ends up in `while(1)` (which actually just waits for deep sleep to happen), but in 5.1 there is a check and it might actually return.

Pushed a commit that skip turn of unreachable code for this function, so the panic can always be there. Seems cleaner than to check for versions (which might break for a later version).

Also added more comments, hopefully explaining the strangeness better